### ISSUE TITLE
Advanced Observability for H2

### ISSUE THREAD

**Issue by adleong** [linkerd/linkerd#1508]
Continuation of https://github.com/linkerd/linkerd/issues/1507

Update the stats filter from https://github.com/linkerd/linkerd/issues/1507 to accept an H2ResponseClassifier as a stack param.  H2ResponseClassifier is essentially the type:

```
PartialFunction[(Request, Try[(Response, Try[Frame])]), ResponseClass]
```

Whenever any of the following conditions are met, the stats filter should consult the classifier and increment the `success` or `failures` stat accordingly:
* The response future fails.  `(req, Throw(e))` should be passed to the classifier.
* The response stream fails.  `(req, Return(rsp, Throw(e)))` should be passed to the classifier.
* The response stream completes successfully.  `(req, Return(rsp, Return(f)))` should be passed to the classifier where `f` is the final frame (EOS) of the response stream.

**Comment by adleong** [comment]
An added wrinkle here is that `onEnd` is a `Future[Unit]` which means we don't have access to the final frame of the stream.  Changing the type of `onEnd` to `Future[Frame]` is probably not a good solution either because this would cause the Stream to hold a reference to the final frame, even after the frame is released, which is a violation of flow control.

I recommend creating a `StreamProxy` class that delegates all calls to an underlying stream but also takes a function `onFrame: Try[Frame] => ()` which is called on each read.  By wrapping the response stream in a StreamProxy, we can give it an `onFrame` that increments the appropriate stats on the final frame.

pseudocode would be something like
```
svc(req).transform {
  case Return(rsp0) =>
    val stream1 = new StreamProxy(rsp0.stream) {
      case Return(f) if f.isEnd => // consult classifier and incr stat
      case Return(f) => // do nothing
      case Throw(e) => // consult classifier and incr stat
    }
    Future.value(rsp0.copy(stream = stream1))
  case Throw(e) => 
    // consult classifier and incr stat
    Future.exception(e)
}
```

**Comment by adleong** [comment]
We should also add some reasonable built-in classifiers:
* Successful iff response and response stream are successful
* Successful if response and response stream are successful and response status is < 500

**Comment by hawkw** [comment]
It occurs to me that with the addition of an `onFrame` function, we also have the ability to collect some pretty granular metrics, such as, say, the number of frames per request, . Is this data we're interested in collecting @adleong?

**Comment by adleong** [comment]
That would be "hella dope".  Similarly: histogram of frame sizes.

**Comment by hawkw** [comment]
Consider it done! :D

**Comment by hawkw** [comment]
Closed b81efb74a48aecda7a064af20aec1362a6e7b6f7
